lib/ASTGen/Sources/ASTGen/DeclAttrs.swift

```diff
@@ -17,12 +17,37 @@ import SwiftIfConfig
 
 @_spi(ExperimentalLanguageFeatures) @_spi(RawSyntax) import SwiftSyntax
 
+/// Calls `visitor` on each declaration that an `@abi` attribute `decl` ought to apply to.
+/// This is usually just `decl` itself, but for pattern binding decls, it's all of the
+/// variables bound to it.
+func visitABIDecls(for decl: BridgedDecl, visitor: (BridgedDecl) -> Void) {
+  if let pbd = BridgedPatternBindingDecl(decl.asPatternBindingDecl) {
+    for pattern in pbd.patterns {
+      for varDecl in pattern.varDecls {
+        visitABIDecls(for: varDecl.asDecl, visitor: visitor)
+      }
+    }
+    return
+  }
+
+  visitor(decl)
+}
+
 extension ASTGenVisitor {
   struct DeclAttributesResult {
     var attributes: BridgedDeclAttributes
     var staticSpelling: BridgedStaticSpelling
     var staticLoc: BridgedSourceLoc
     var initContext: BridgedPatternBindingInitializer?
+
+    func attach(to decl: BridgedDecl) {
+      decl.attrs = attributes
+      if let abiAttr = attributes.attrs.lazy.map(\.asABIAttr).compactMap(BridgedABIAttr.init).first {
+        visitABIDecls(for: decl) { declToConnect in
+          abiAttr.connectToInverse(attachedTo: declToConnect)
+        }
+      }
```

Comment:
Since this is a duplicated logic between C++ parser and ASTGen, is it possible to do this in AST? For example, make `Decl::setParsedAttrs(const DeclAttributes &)` and use it from C++ parser and ASTGen.

TBH, I'm not a fan of this API. This feels like `element.append(to: array)`